cmd/kubeless/function.go

```diff
@@ -117,41 +119,72 @@ func getFileSha256(file string) (string, error) {
 	return checksum, err
 }
 
-func uploadFunction(file string, cli kubernetes.Interface) (string, string, string, error) {
-	var function, contentType, checksum string
-	stats, err := os.Stat(file)
+func isMinioAvailable(cli kubernetes.Interface) bool {
+	_, err := cli.Core().Services("kubeless").Get("minio", metav1.GetOptions{})
 	if err != nil {
-		return "", "", "", err
+		return false
 	}
-	if stats.Size() > int64(50*1024*1024) { // TODO: Make the max file size (50 MB) configurable
-		err = errors.New("The maximum size of a function is 50MB")
-		return "", "", "", err
-	}
-	// If an object storage service is not available check
-	// that the file is not over 1MB to store it as a Custom Resource
-	if stats.Size() > int64(1*1024*1024) {
-		err = errors.New("Unable to deploy functions over 1MB withouth a storage service")
-		return "", "", "", err
+	minioPods, err := cli.Core().Pods("kubeless").List(metav1.ListOptions{
+		LabelSelector: "kubeless=minio",
+	})
+	for i := range minioPods.Items {
+		if minioPods.Items[i].Status.Phase != "Running" {
+			logrus.Warn("Found unhealthy Minio pod, disabling upload")
+			return false
+		}
 	}
-	functionBytes, err := ioutil.ReadFile(file)
+	return true
+}
+
+func uploadFunction(file string, cli kubernetes.Interface) (string, string, string, error) {
+	var function, contentType, checksum string
+	stats, err := os.Stat(file)
 	if err != nil {
 		return "", "", "", err
 	}
+	rawChecksum, err := getFileSha256(file)
 	if err != nil {
 		return "", "", "", err
 	}
-	fileType := http.DetectContentType(functionBytes)
-	if strings.Contains(fileType, "text/plain") {
-		function = string(functionBytes[:])
-		contentType = "text"
+	checksum = "sha256:" + rawChecksum
+
+	if isMinioAvailable(cli) {
+		if err != nil {
+			return "", "", "", err
+		}
+		function, err = minio.UploadFunction(file, rawChecksum, stats, cli)
+		if err != nil {
+			return "", "", "", err
+		}
+		contentType = "URL"
+		if err != nil {
+			return "", "", "", err
+		}
 	} else {
-		function = base64.StdEncoding.EncodeToString(functionBytes)
-		contentType = "base64"
-	}
-	c, err := getFileSha256(file)
-	checksum = "sha256:" + c
-	if err != nil {
-		return "", "", "", err
+		// If an object storage service is not available check
+		// that the file is not over 1MB to store it as a Custom Resource
+		if stats.Size() > int64(1*1024*1024) {
+			err = errors.New("Unable to deploy functions over 1MB withouth a storage service")
+			return "", "", "", err
+		}
+		functionBytes, err := ioutil.ReadFile(file)
+		if err != nil {
+			return "", "", "", err
+		}
+		if err != nil {
+			return "", "", "", err
+		}
+		fileType := http.DetectContentType(functionBytes)
```

Comment:
The check should really be "can be encoded as utf8 (a json string)" - ie `if utf8.Valid(functionBytes) {...}`